_drafts/2023-06-01-functional-error-handling-in-kotlin-part-2.md

```diff
@@ -0,0 +1,912 @@
+
+```kotlin
+class JobService(private val jobs: Jobs) {
+
+    fun printOptionJob(jobId: JobId) {
+        val maybeJob: Result<Job?> = jobs.findById(jobId)
+        if (maybeJob.isSuccess) {
+            maybeJob.getOrNull()?.apply { println("Job found: $this") } ?: println("Job not found for id $jobId")
+        } else {
+            println("Something went wrong: ${maybeJob.exceptionOrNull()}")
+        }
+    }
+}
+```
+
+The above code also introduces some of the extractor methods defined on the `Result` type:
+
+* `getOrNull` returns the value of the `Result` or `null` otherwise.
+* `getOrThrow` returns the value of the `Result` or the throw of the exception contained in the `Result.Failure` instance.
+* `getOrDefault` returns the value of the `Result` or a given default value if the `Result` is a failure.
+* `getOrElse` returns the value of the `Result`, or the result of a given lambda, `onFailure: (exception: Throwable) -> R` if the `Result` is a failure.
+
+Conversely, the method `exceptionOrNull` returns the exception in the `Result.Failure` instance or `null` otherwise.
+
+If we execute the above method using a valid `JobId`, we'll get the following output:
+
+```text
+Job found: Job(id=JobId(value=1), company=Company(name=Apple, Inc.), role=Role(name=Software Engineer), salary=Salary(value=70000.0))
+```
+
+However, it's far more idiomatic to transform and react to values of a `Result` rather than extracting it. To apply a transformation to the value of a `Result`, we can use the classic `map` function:
+
+```kotlin
+val appleJobSalary: Result<Salary> = appleJob.map { it.salary }
+```
+
+Instead, we can use the `mapCatching` function if the transformation to apply to the value of the `Result` can throw an exception. Let's reuse our `CurrencyConverter` class from the first part of the series. The `CurrencyConverter` is a class that converts amounts from one currency to another one. In detail, we want to develop a method to convert an amount in USD to EUR. This time, we add validation on the input amount:
+
+```kotlin
+class CurrencyConverter {
+    @Throws(IllegalArgumentException::class)
+    fun convertUsdToEur(amount: Double?): Double =
+        if (amount != null && amount >= 0.0) {
+            amount * 0.91
+        } else {
+            throw IllegalArgumentException("Amount must be positive")
+        }
+}
+```
+
+The converter will throw an exception if the input amount is `null` or negative. Let's think about the converter as an external library that we can't change. We can use the `mapCatching` function to convert the salary of a job from USD to EUR, handling the fact that the conversion can fail, throwing an exception:
+
+```kotlin
+fun getSalaryInEur(jobId: JobId): Result<Double> =
```

Comment:
this should stay in JobService, which now needs a constructor argument to CurrencyConverter